src/libcollections/btree/map.rs

```diff
@@ -463,6 +464,193 @@ impl<K: Ord, V> BTreeMap<K, V> {
+    pub fn append(&mut self, other: &mut Self) {
+        // Do we have to append anything at all?
+        if other.len() == 0 {
+            return;
+        }
+
+        // If the values of `b` of `self` and `other` are equal, we can just swap them if `self` is
+        // empty.
+        if self.len() == 0 && self.b == other.b {
+            mem::swap(self, other);
+        }
+
+        // First, we merge `self` and `other` into a sorted sequence in linear time.
+        let self_b = self.b;
+        let other_b = other.b;
+        let mut self_iter = mem::replace(self, BTreeMap::with_b(self_b)).into_iter();
+        let mut other_iter = mem::replace(other, BTreeMap::with_b(other_b)).into_iter();
+
+        // Consume `self_iter` and `other_iter` and push elements into `elements`.
+        let mut elements = Vec::new();
+        let mut self_cur = self_iter.next();
+        let mut other_cur = other_iter.next();
+
+        loop {
+            match (self_cur, other_cur) {
+                (Some(self_element), Some(other_element)) => {
+                    // Check which elements comes first and only advance the corresponding iterator.
+                    // If two keys are equal, take the value from `other`.
+                    if self_element.0 < other_element.0 {
+                        elements.push(self_element);
+                        self_cur = self_iter.next();
+                        other_cur = Some(other_element);
+                    } else if self_element.0 == other_element.0 {
+                        elements.push(other_element);
+                        self_cur = self_iter.next();
+                        other_cur = other_iter.next();
+                    } else {
+                        elements.push(other_element);
+                        self_cur = Some(self_element);
+                        other_cur = other_iter.next();
+                    }
+                },
+                (None, Some(other_element)) => {
+                    // `self_iter` is exhausted, push all elements from `other` into `elements`.
+                    elements.push(other_element);
+                    for element in other_iter {
+                        elements.push(element);
+                    }
+                    break;
+                },
+                (Some(self_element), None) => {
+                    // `other_iter` is exhausted, push all elements from `self` into `elements`.
+                    elements.push(self_element);
+                    for element in self_iter {
+                        elements.push(element);
+                    }
+                    break;
+                },
+                (None, None) => break,
+            }
+        }
+
+        // Second, we build a tree from the sorted sequence in linear time.
+        self.length = elements.len();
+        let (depth, root) = Node::from_sorted_iter(elements.into_iter(), self.length, self_b);
```

Comment:
Bleh, I hadn't though of the equal key case. It still might be possible and not too difficult, as in the worst case, only the "right edge" of the BTree will be left underfull, and every underfull node except for the root (which, if I remember correctly, is allowed to be underfull) is adjacent to a completely full node, allowing a simple steal to fix things.